File: day6/lanternfish.py
#!/usr/bin/env python3

from collections import Counter
from collections import deque

def main(filename, days):
  inputfile = open(filename, 'r')
  line = inputfile.readline()
  starting_fish = [int(i) for i in line.strip().split(',')]

  # Count the number of fish with each timer value.
  fish_counter = Counter()
  fish_counter.update(starting_fish)

  # Queue of the counts of the number of fish with each timer value.
  # The leftmost items have a timer of zero, the next 1, etc.
  # The rightmost items have a timer of 6.
  existing_fish = deque()
  # Queue of the counts of the number of *new* fish with each timer value.
  # The leftmost items have a timer of 7, the next have a timer of 8.
  new_fish = deque()

  # First 7 (timers of 0-6) go in the existing_fish queue.
  for i in range(7):
    existing_fish.append(fish_counter[i])
  # Timers of 7 and 8 go in the new_fish queue.
  for i in range(7, 9):
    new_fish.append(fish_counter[i])
  
  for day in range(1, days + 1):
    # Remove the counts for fish with timer 0 and 7 from the queues.
    # This has the side-effect of promoting all the other counts 
    multipliers = existing_fish.popleft()
    graduators = new_fish.popleft()
    # Each fish with timer 0 spawns that many new fish (timer 8)
    new_fish.append(multipliers)
    # Also add them to the back of the queue (timer 6) along with the
    # ones that graduated from the new_fish queue (were timer 7).
    existing_fish.append(graduators + multipliers)
    # print('After day', day, 'state is:', list(existing_fish) + list(new_fish))

  result = str(sum(existing_fish) + sum(new_fish))
  if len(result) < 100:
    print(filename, days, 'days, total number of fish:', result)
  else:
    print(filename, days, 'days, number of digits in total number of fish:', len(result))


if __name__ == '__main__':
  main('example_input.txt', 80)
  main('input.txt', 80)
  main('example_input.txt', 256)
  main('input.txt', 256)
  main('example_input.txt', 9999999)

File: day6/lanternfish.cc

#include <iostream>
#include <fstream>
#include <gmpxx.h>
#include <string>
#include <deque>
#include <sstream>
#include <memory>

void run(const std::string& filename, int days) {
  std::ifstream infile;
  infile.open(filename);
  std::deque<std::unique_ptr<mpz_class>> fish;
  for (int i = 0; i < 9; i++) {
    fish.emplace_back(std::make_unique<mpz_class>(0));
  }
  std::string v;
  while (std::getline(infile, v, ',')) {
    (*fish[std::stoi(v)])++;
  }

  for (int day = 0; day < days; day++) {
    std::unique_ptr<mpz_class> multipliers = std::move(fish.front());
    fish.pop_front();
    (*fish[6]) += *multipliers;
    fish.emplace_back(std::move(multipliers));
    // std::cout << "After day " << day << " state is: "
    //   << fish[0] << ", " << fish[1] << ", " << fish[2] << ", "
    //   << fish[3] << ", " << fish[4] << ", " << fish[5] << ", "
    //   << fish[6] << ", " << fish[7] << ", " << fish[8] << std::endl;
  }

  mpz_class result = 0;
  for (const std::unique_ptr<mpz_class>& count : fish) {
    result += *count;
  }
  std::string result_string = result.get_str();
  if (result_string.size() < 100) {
    std::cout << filename << " after " << days
      << " days the number of fish is: " << result_string << std::endl;
  } else {
    std::cout << filename << " after " << days
      << " days the number of digits in the number of fish is: "
      << result_string.size() << std::endl;
  }
}

int main() {
    run("example_input.txt", 80);
    run("input.txt", 80);
    run("example_input.txt", 256);
    run("input.txt", 256);
    run("example_input.txt", 9999999);
    return 0;
}
